lightning/src/routing/router.rs

```diff
@@ -223,77 +374,170 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
+			// It may be misleading because we might later choose to reduce the value transferred
+			// over these channels, and the channel which was insufficient might become sufficient.
+			// Worst case: we drop a good channel here because it can't cover the high following fees
+			// caused by one expensive channel, but then this channel could have been used if the amount being
+			// transferred over this path is lower.
+			// We do this for now, but this check is a subject for removal.
+			let can_cover_following_hops = *available_liquidity_msat > $min_total_fee_to_payee;
+
+			// Includes paying fees for the use of the following channels.
+			let amount_to_transfer_over_msat: u64 = cmp::min(*available_liquidity_msat, value_left_to_collect_msat);
+
 			//TODO: Explore simply adding fee to hit htlc_minimum_msat
-			if $starting_fee_msat as u64 + final_value_msat >= $directional_info.htlc_minimum_msat {
-				let proportional_fee_millions = ($starting_fee_msat + final_value_msat).checked_mul($directional_info.fees.proportional_millionths as u64);
-				if let Some(new_fee) = proportional_fee_millions.and_then(|part| {
-						($directional_info.fees.base_msat as u64).checked_add(part / 1000000) })
-				{
-					let mut total_fee = $starting_fee_msat as u64;
-					let hm_entry = dist.entry(&$src_node_id);
-					let old_entry = hm_entry.or_insert_with(|| {
-						let mut fee_base_msat = u32::max_value();
-						let mut fee_proportional_millionths = u32::max_value();
-						if let Some(fees) = network.get_nodes().get(&$src_node_id).and_then(|node| node.lowest_inbound_channel_fees) {
-							fee_base_msat = fees.base_msat;
-							fee_proportional_millionths = fees.proportional_millionths;
-						}
-						(u64::max_value(),
-							fee_base_msat,
-							fee_proportional_millionths,
-							RouteHop {
-								pubkey: $dest_node_id.clone(),
-								node_features: NodeFeatures::empty(),
-								short_channel_id: 0,
-								channel_features: $chan_features.clone(),
-								fee_msat: 0,
-								cltv_expiry_delta: 0,
-							},
-						)
-					});
+			if has_sufficient_liquidity && can_cover_following_hops && amount_to_transfer_over_msat >= $directional_info.htlc_minimum_msat {
+				let hm_entry = dist.entry(&$src_node_id);
+				let old_entry = hm_entry.or_insert_with(|| {
+					// If there was previously no known way to access the source node (recall it goes payee-to-payer) of $chan_id,
+					// first add a semi-dummy record just to compute the fees to reach the source node.
+					// This will affect our decision on selecting $chan_id as a way to reach the $dest_node_id.
+					let mut fee_base_msat = u32::max_value();
+					let mut fee_proportional_millionths = u32::max_value();
+					if let Some(Some(fees)) = network.get_nodes().get(&$src_node_id).map(|node| node.lowest_inbound_channel_fees) {
+						fee_base_msat = fees.base_msat;
+						fee_proportional_millionths = fees.proportional_millionths;
```

Comment:
Is it a problem that these two fields may refer to different channels, and/or may be the fee schema of $chan_id (iiuc)?

Reply:
I don't think so, it's just used in a heuristic. I mean, yeah, it may be gameable a bit, but I don't think we should be concerned about it at this point.